File: procode_import/fix_quiz_code_snippets.js
// fix_quiz_code_snippets.js
// This script updates quiz questions that are missing their code snippets

const admin = require("firebase-admin");
const serviceAccount = require("./serviceAccountKey.json");

// Initialize Firebase Admin
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount),
});

const db = admin.firestore();

// Map of quiz titles to their questions with code snippets
const quizCodeSnippets = {
  "Python Variables & Data Types": {
    "What will be the output?": {
      codeSnippet: "x = 5\ny = '5'\nprint(x == y)",
      correctAnswer: "False",
    },
    "What is the result of: 10 // 3": {
      codeSnippet: "10 // 3",
      correctAnswer: "3",
    },
    "What is the output?": {
      codeSnippet: "x = [1, 2, 3]\ny = x\ny.append(4)\nprint(x)",
      correctAnswer: "[1, 2, 3, 4]",
    },
  },
  "Python Control Flow": {
    "What will this code print?": {
      codeSnippet: "for i in range(3):\n    print(i)",
      correctAnswer: "0 1 2",
    },
    "What is the output?": [
      {
        questionText:
          "x = 5\nif x > 10:\n    print('A')\nelif x > 3:\n    print('B')\nelse:\n    print('C')",
        codeSnippet:
          "x = 5\nif x > 10:\n    print('A')\nelif x > 3:\n    print('B')\nelse:\n    print('C')",
        correctAnswer: "B",
      },
      {
        questionText: "for i in range(2, 8, 2):\n    print(i, end=' ')",
        codeSnippet: "for i in range(2, 8, 2):\n    print(i, end=' ')",
        correctAnswer: "2 4 6",
      },
    ],
    "What does this list comprehension produce?": {
      codeSnippet: "[x**2 for x in range(4) if x % 2 == 0]",
      correctAnswer: "[0, 4]",
    },
  },
  "Python Master Challenge - Week 1": {
    "What is the output of this code?": [
      {
        questionText: "x = [1, 2, 3]\ny = x[:]\ny.append(4)\nprint(x)",
        codeSnippet: "x = [1, 2, 3]\ny = x[:]\ny.append(4)\nprint(x)",
        correctAnswer: "[1, 2, 3]",
      },
      {
        questionText:
          "def func(a, b=2, c=3):\n    return a + b + c\n\nprint(func(1, c=4))",
        codeSnippet:
          "def func(a, b=2, c=3):\n    return a + b + c\n\nprint(func(1, c=4))",
        correctAnswer: "7",
      },
      {
        questionText: "a = [1, 2, 3]\nb = a\na = a + [4]\nprint(b)",
        codeSnippet: "a = [1, 2, 3]\nb = a\na = a + [4]\nprint(b)",
        correctAnswer: "[1, 2, 3]",
      },
      {
        questionText:
          "def outer():\n    x = 1\n    def inner():\n        nonlocal x\n        x = 2\n    inner()\n    return x\n\nprint(outer())",
        codeSnippet:
          "def outer():\n    x = 1\n    def inner():\n        nonlocal x\n        x = 2\n    inner()\n    return x\n\nprint(outer())",
        correctAnswer: "2",
      },
    ],
    "What is the result of: 'Python'[1:4]": {
      codeSnippet: "'Python'[1:4]",
      correctAnswer: "'yth'",
    },
  },
};

async function fixCodeSnippets() {
  console.log("🔧 Starting to fix code snippets in quiz questions...\n");

  try {
    // Get all quizzes
    const quizzesSnapshot = await db.collection("quizzes").get();
    console.log(`Found ${quizzesSnapshot.size} quizzes\n`);

    let totalFixed = 0;

    for (const quizDoc of quizzesSnapshot.docs) {
      const quizData = quizDoc.data();
      const quizTitle = quizData.title;

      if (!quizCodeSnippets[quizTitle]) {
        console.log(
          `⏭️  Skipping quiz: ${quizTitle} (no code snippets needed)`
        );
        continue;
      }

      console.log(`\n📝 Processing quiz: ${quizTitle}`);

      // Get all questions for this quiz
      const questionsSnapshot = await db
        .collection("quizzes")
        .doc(quizDoc.id)
        .collection("questions")
        .get();

      const batch = db.batch();
      let fixedInQuiz = 0;

      for (const questionDoc of questionsSnapshot.docs) {
        const questionData = questionDoc.data();
        const questionText = questionData.question;
        const snippetInfo = quizCodeSnippets[quizTitle][questionText];

        if (snippetInfo) {
          let codeSnippet = null;

          // Handle array of questions with same text
          if (Array.isArray(snippetInfo)) {
            // Find matching question by checking if the question contains the code
            const match = snippetInfo.find(
              (info) =>
                questionText.includes(info.questionText.substring(0, 20)) ||
                questionData.correctAnswer === info.correctAnswer
            );
            if (match) {
              codeSnippet = match.codeSnippet;
            }
          } else {
            codeSnippet = snippetInfo.codeSnippet;
          }

          if (
            codeSnippet &&
            (!questionData.codeSnippet || questionData.codeSnippet === "")
          ) {
            console.log(
              `  ✅ Adding code snippet to: "${questionText.substring(
                0,
                50
              )}..."`
            );

            const questionRef = db
              .collection("quizzes")
              .doc(quizDoc.id)
              .collection("questions")
              .doc(questionDoc.id);

            batch.update(questionRef, { codeSnippet: codeSnippet });
            fixedInQuiz++;
          }
        }
      }

      if (fixedInQuiz > 0) {
        await batch.commit();
        console.log(`  ✅ Fixed ${fixedInQuiz} questions in this quiz`);
        totalFixed += fixedInQuiz;
      } else {
        console.log(`  ℹ️  No questions needed fixing in this quiz`);
      }
    }

    console.log(`\n✅ Finished! Fixed ${totalFixed} questions total.`);
  } catch (error) {
    console.error("❌ Error fixing code snippets:", error);
  } finally {
    process.exit();
  }
}

// Run the fix
fixCodeSnippets();

File: lib/utils/fix_quiz_code_snippets.dart
// lib/utils/fix_quiz_code_snippets.dart
// Run this from your Flutter app to fix missing code snippets

import 'package:cloud_firestore/cloud_firestore.dart';

/// Utility class to fix missing code snippets in quiz questions
/// This was created to retroactively add code snippets to existing quizzes
/// Some quiz questions reference code but were missing the actual snippets
class QuizCodeSnippetFixer {
  static final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  /// Main method to fix all missing code snippets across all quizzes
  /// Call this from a button in your app (admin/debug screen)
  static Future<void> fixMissingCodeSnippets() async {
    print('Starting to fix missing code snippets...');

    try {
      // Get all quizzes from Firestore
      final quizzesSnapshot = await _firestore.collection('quizzes').get();

      int totalFixed = 0;

      // Process each quiz
      for (final quizDoc in quizzesSnapshot.docs) {
        final quizData = quizDoc.data();
        final quizTitle = quizData['title'];

        print('Processing quiz: $quizTitle');

        // Get all questions for this quiz
        final questionsSnapshot = await _firestore
            .collection('quizzes')
            .doc(quizDoc.id)
            .collection('questions')
            .get();

        // Use batch write for efficiency
        final batch = _firestore.batch();
        int fixedInQuiz = 0;

        // Check each question
        for (final questionDoc in questionsSnapshot.docs) {
          final questionData = questionDoc.data();
          final question = questionData['question'];
          final existingSnippet = questionData['codeSnippet'];

          // Determine if this question needs a code snippet
          String? codeSnippet;

          // Python Variables & Data Types quiz
          // These questions reference code but were missing snippets
          if (quizTitle == "Python Variables & Data Types") {
            if (question == "What will be the output?" &&
                questionData['options']?.contains('True')) {
              // Type comparison question
              codeSnippet = "x = 5\ny = '5'\nprint(x == y)";
            } else if (question.contains("10 // 3")) {
              // Floor division question
              codeSnippet = "10 // 3";
            } else if (question == "What is the output?" &&
                questionData['options']?.contains('[1, 2, 3, 4]')) {
              // List reference question
              codeSnippet = "x = [1, 2, 3]\ny = x\ny.append(4)\nprint(x)";
            } else if (question.contains("type(3.14)")) {
              // Type checking question
              codeSnippet = "type(3.14)";
            }
          }

          // Python Control Flow quiz
          else if (quizTitle == "Python Control Flow") {
            if (question.contains("range(3)")) {
              // Range function demonstration
              codeSnippet = "for i in range(3):\n    print(i)";
            } else if (question == "What is the output?" &&
                questionData['options']?.contains('B')) {
              // If-elif-else logic
              codeSnippet =
                  "x = 5\nif x > 10:\n    print('A')\nelif x > 3:\n    print('B')\nelse:\n    print('C')";
            } else if (question == "What is the output?" &&
                questionData['options']?.contains('2 4 6')) {
              // Range with step parameter
              codeSnippet = "for i in range(2, 8, 2):\n    print(i, end=' ')";
            } else if (question.contains("list comprehension")) {
              // List comprehension example
              codeSnippet = "[x**2 for x in range(4) if x % 2 == 0]";
            }
          }

          // Python Master Challenge - advanced concepts
          else if (quizTitle.contains("Python Master Challenge")) {
            if (question == "What is the output of this code?" &&
                questionData['options']?.contains('[1, 2, 3]')) {
              // List slicing and copying
              codeSnippet = "x = [1, 2, 3]\ny = x[:]\ny.append(4)\nprint(x)";
            } else if (question == "What is the output?" &&
                questionData['options']?.contains('7')) {
              // Function default parameters
              codeSnippet =
                  "def func(a, b=2, c=3):\n    return a + b + c\n\nprint(func(1, c=4))";
            } else if (question.contains("'Python'[1:4]")) {
              // String slicing
              codeSnippet = "'Python'[1:4]";
            } else if (question == "What is the output?" &&
                questionData['options']?.contains('[1, 2, 3]') &&
                questionData['correctAnswer'] == '[1, 2, 3]') {
              // List assignment behavior
              codeSnippet = "a = [1, 2, 3]\nb = a\na = a + [4]\nprint(b)";
            } else if (question == "What is the output?" &&
                questionData['correctAnswer'] == '2') {
              // Nonlocal variable scope
              codeSnippet =
                  "def outer():\n    x = 1\n    def inner():\n        nonlocal x\n        x = 2\n    inner()\n    return x\n\nprint(outer())";
            }
          }

          // Only update if we found a snippet and it's currently missing
          if (codeSnippet != null &&
              (existingSnippet == null || existingSnippet == '')) {
            print('  Adding code snippet to: ${question.substring(0, 50)}...');

            batch.update(
              questionDoc.reference,
              {'codeSnippet': codeSnippet},
            );
            fixedInQuiz++;
          }
        }

        // Commit batch if we made any changes
        if (fixedInQuiz > 0) {
          await batch.commit();
          print('  Fixed $fixedInQuiz questions in $quizTitle');
          totalFixed += fixedInQuiz;
        }
      }

      print('Finished! Fixed $totalFixed questions total.');
    } catch (e) {
      print('Error fixing code snippets: $e');
      rethrow;
    }
  }

  /// Quick fix method for a specific quiz by ID
  /// Used when you know exactly which quiz needs fixing
  static Future<void> fixSpecificQuiz(String quizId) async {
    try {
      // Get questions ordered by index
      final questionsSnapshot = await _firestore
          .collection('quizzes')
          .doc(quizId)
          .collection('questions')
          .orderBy('orderIndex')
          .get();

      final batch = _firestore.batch();

      // Map of question indices to code snippets
      // This is for Python Variables quiz specifically
      final codeSnippets = {
        1: "type(3.14)", // What is the output of: type(3.14)
        4: "x = 5\ny = '5'\nprint(x == y)", // What will be the output?
        6: "10 // 3", // What is the result of: 10 // 3
        9: "x = [1, 2, 3]\ny = x\ny.append(4)\nprint(x)", // What is the output?
      };

      // Apply fixes based on question index
      for (int i = 0; i < questionsSnapshot.docs.length; i++) {
        final doc = questionsSnapshot.docs[i];
        final codeSnippet = codeSnippets[i];

        if (codeSnippet != null) {
          batch.update(doc.reference, {'codeSnippet': codeSnippet});
        }
      }

      await batch.commit();
      print('Fixed code snippets for quiz');
    } catch (e) {
      print('Error: $e');
    }
  }
}

// Example usage - Add this to a temporary button in admin/debug screen:
// ElevatedButton(
//   onPressed: () async {
//     try {
//       await QuizCodeSnippetFixer.fixMissingCodeSnippets();
//       ScaffoldMessenger.of(context).showSnackBar(
//         SnackBar(content: Text('Code snippets fixed!')),
//       );
//     } catch (e) {
//       ScaffoldMessenger.of(context).showSnackBar(
//         SnackBar(content: Text('Error: $e')),
//       );
//     }
//   },
//   child: Text('Fix Quiz Code Snippets'),
// ),
